Edit geoms (points, polylines, polygons, rectangle)

vladyslav-tk opened the pull request.

gberaudo commented on src/toolbox/AreaOfInterestDrawer.js:
```diff
@@ -391,6 +405,93 @@ class NgmAreaOfInterestDrawer extends I18nMixin(LitElement) {
+    const entity = this.interestAreasDataSource.entities.getById(id);
+    const type = entity.properties.type.getValue();
+    if (!entity.isShowing) {
+      entity.show = !entity.isShowing;
```
```suggestion
      entity.show = true;
```

vladyslav-tk commented on src/draw/CesiumDraw.js:
```diff
@@ -68,9 +72,29 @@ export class CesiumDraw extends EventTarget {
+        if (this.entityForEdit) {
+          this.eventHandler_.setInputAction(event => this.onLeftDown_(event), ScreenSpaceEventType.LEFT_DOWN);
+          this.eventHandler_.setInputAction(event => this.onLeftUp_(event), ScreenSpaceEventType.LEFT_UP);
+          if (this.type !== 'point') {
```
but there is only one case - not point

gberaudo replied:
Your code handles all geometries:
- points (in the else block),
- polygons (including rectangles) with `this.entityForEdit.polygon` and `this.type === 'rectangle'`
- line in the second case of the ternary: `this.entityForEdit.polyline`.

It would help maintenance to have each case clearly defined.

vladyslav-tk commented on src/draw/CesiumDraw.js:
```diff
@@ -68,9 +72,29 @@ export class CesiumDraw extends EventTarget {
+              this.entityForEdit.polygon.hierarchy.getValue().positions :
+              this.entityForEdit.polyline.positions.getValue();
+            if (this.type === 'rectangle') {
+              positions = positions.slice(0, 3);
```
no for rectangle, always 4

gberaudo replied:
ok, then you can remove the parameters.

vladyslav-tk commented on src/draw/CesiumDraw.js:
```diff
@@ -269,12 +306,33 @@ export class CesiumDraw extends EventTarget {
+          this.sketchPoint_.position = position;
+          this.activePoints_[this.sketchPoint_.properties.index] = position;
+          if (this.type === 'line') {
+            this.entityForEdit.polyline.positions = this.activePoints_;
```
yes, this line updates polyline positions

vladyslav-tk commented on src/draw/CesiumDraw.js:
```diff
@@ -269,12 +306,33 @@ export class CesiumDraw extends EventTarget {
+            const hierarchy = this.entityForEdit.polygon.hierarchy.getValue();
+            let positions = this.activePoints_;
+            if (this.type === 'rectangle') {
+              positions = rectanglify(this.activePoints_).map(p => Cartesian3.clone(p));
```
yes, to avoid issues with future drawing

gberaudo replied:
Can we instead clone the points when starting a new drawing?

vladyslav-tk commented on src/draw/CesiumDraw.js:
```diff
@@ -269,12 +306,33 @@ export class CesiumDraw extends EventTarget {
+                sp.position = positions[key];
+              });
+            }
+            this.entityForEdit.polygon.hierarchy = {...hierarchy, positions};
```
holes

gberaudo replied:
Is it possible to draw a polygon with holes?

vladyslav-tk commented on src/draw/CesiumDraw.js:
```diff
@@ -284,6 +342,33 @@ export class CesiumDraw extends EventTarget {
+  onLeftDown_(event) {
+    this.leftPressed = true;
+    if (this.entityForEdit) {
+      const objects = this.viewer_.scene.drillPick(event.position, 5, 5, 5);
```
for better picking of sketch points

vladyslav-tk commented:
@gberaudo I made updates according to your comments and added answers.

I have questions about the general comment:

 > I would have expected the interactions code to be reusable in other projects and the code for handling each geometry type to be more clearly separated.

'interactions code' you mean `CesiumDraw` changes? It stays reusable as before, there is only one new property is the entity that needs to be edited, no dependencies.

> code for handling each geometry type to be more clearly separated

What do you mean? I thought there are no sens to separate as currently, it is few lines of code for each geometry

gberaudo commented on src/draw/CesiumDraw.js:
```diff
@@ -68,9 +72,29 @@ export class CesiumDraw extends EventTarget {
+              this.entityForEdit.polygon.hierarchy.getValue().positions :
+              this.entityForEdit.polyline.positions.getValue();
+            if (this.type === 'rectangle') {
+              positions = positions.slice(0, 3);
```
```suggestion
              positions = positions.slice();
```

gberaudo commented:
Sorry, I missed your questions.

> 'interactions code' you mean CesiumDraw changes? It stays reusable as before, there is only one > new property is the entity that needs to be edited, no dependencies.

Right, it is all good. Read the PR too quickly. Thanks.

>>    code for handling each geometry type to be more clearly separated

> What do you mean? I thought there are no sens to separate as currently, it is few lines of code for > each geometry

Using a switch is only an example about how to make the code more readable.
I think that testing `this.type` will already quite improve readability.

vladyslav-tk commented:
@gberaudo I made updates:

> Your code handles all geometries

I moved edit activation to separate function and add switch

> ok, then you can remove the parameters.

`slice` without params returns the same array, I replaced it with `pop`

> Is it possible to draw a polygon with holes?

Currently no but I thought if we add it in future we will search for a place where it removes. I removed it

> Can we instead clone the points when starting a new drawing?

The reason was in `rectanglify` function, somehow during drawing, it breaks the previously edited rectangle (`activePoints ` cleanups each time).  I added clone to the variables where it is necessary

gberaudo commented on src/draw/CesiumDraw.js:
```diff
@@ -144,16 +184,24 @@ export class CesiumDraw extends EventTarget {
 
-  drawSketchPoint_(position) {
-    return this.viewer_.entities.add({
+  drawSketchPoint_(position, edit = false) {
```
The function could be named better.
What about calling it `createSketchPoint` and adding it to the entity collection in the calling code?
Also, can you pass an option object instead of a boolean? `{edit: true}`.

vladyslav-tk commented:
I will merge this PR and implement improvements in 3D volume PR if have time
